src/H5.c

```diff
@@ -346,108 +350,137 @@ H5_term_library(void)
+    ,   TERMINATOR(M, false)
+    ,   TERMINATOR(S, false)
+    ,   TERMINATOR(T, false)
+        /* Wait to shut down low-level packages like AC until after
+         * the preceding high-level packages have shut down.  This prevents
+         * low-level objects from closing "out from underneath" their
+         * reliant high-level objects.
          */
-        if (pending == 0) {
-            pending += DOWN(AC);
-            /* Shut down the "pluggable" interfaces, before the plugin framework */
-            pending += DOWN(Z);
-            pending += DOWN(FD);
-            pending += DOWN(VL);
-            /* Don't shut down the plugin code until all "pluggable" interfaces (Z, FD, PL) are shut down */
-            if (pending == 0)
-                pending += DOWN(PL);
-            /* Don't shut down the error code until other APIs which use it are shut down */
-            if (pending == 0)
-                pending += DOWN(E);
-            /* Don't shut down the ID code until other APIs which use them are shut down */
-            if (pending == 0)
-                pending += DOWN(I);
-            /* Don't shut down the skip list code until everything that uses it is down */
-            if (pending == 0)
-                pending += DOWN(SL);
-            /* Don't shut down the free list code until everything that uses it is down */
-            if (pending == 0)
-                pending += DOWN(FL);
-            /* Don't shut down the API context code until _everything_ else is down */
-            if (pending == 0)
-                pending += DOWN(CX);
-        } /* end if */
+    ,   TERMINATOR(AC, true)
+        /* Shut down the "pluggable" interfaces, before the plugin framework */
+    ,   TERMINATOR(Z, false)
+    ,   TERMINATOR(FD, false)
+    ,   TERMINATOR(VL, false)
+        /* Don't shut down the plugin code until all "pluggable" interfaces
+         * (Z, FD, PL) are shut down
+         */
+    ,   TERMINATOR(PL, true)
+        /* Shut down the following packages in strictly the order given
+         * by the table.
+         */
+    ,   TERMINATOR(E, true)
+    ,   TERMINATOR(I, true)
+    ,   TERMINATOR(SL, true)
+    ,   TERMINATOR(FL, true)
+    ,   TERMINATOR(CX, true)
+    };
+
+    do {
+        pending = 0;
+        for (i = 0; i < NELMTS(terminator); i++) {
+            if (terminator[i].completed)
+                continue;
+            if (pending != 0 && terminator[i].await_prior)
+                break;
+            if (terminator[i].func() == 0) {
+                terminator[i].completed = true;
+                continue;
+            }
+
+            /* log a package when its terminator needs to be retried */
+            pending++;
+            nprinted = snprintf(next, nleft, "%s%s",
+                (next != loop) ? "," : "", terminator[i].name);
+            if (nprinted < 0)
+                continue;
+            if ((size_t)nprinted >= nleft)
+                nprinted = snprintf(next, nleft, "...");
```

Comment:
HD prefixes on POSIX and std C calls. Please don't change the coding standard unilaterally.